### PR TITLE
Check App size before update

### PR DESCRIPTION
**Pull request by mike-sul**
- Download a layer manifest if present.
- Get a list of new App's layers that are missing in a local store.
- Check if there is sufficient amount of available storage to accommodate the missing layers of a new App.
- Add tests to cover the use-case happy path as well as edge cases.

### CODE REVIEW

File: src/docker/restorableappengine.cc [hunk 1]
```diff
@@ -299,6 +319,87 @@ void RestorableAppEngine::pullApp(const Uri& uri, const boost::filesystem::path&
...
+  const Json::Value engine_info{docker_client_->engineInfo()};
+  if (!(engine_info.isObject() && engine_info.isMember("Arch"))) {
+    LOG_WARNING << "Failed to get an info about a system architecture";
+    return;
+  }
+
+  const auto arch{engine_info["Arch"].asString()};
+  const auto layers_manifest{manifest.layersManifest(arch)};
+  if (!layers_manifest.isObject()) {
+    LOG_WARNING << "App layers' manifest is missing, skip checking an App update size";
```
Comment (doanac) [inline src/docker/restorableappengine.cc:333]: once compose-publish is updated, this would be a real error, right? i.e. maybe LOG_ERROR here?

**Review by doanac — COMMENTED** [review]
this looks pretty good. nice feature

File: src/docker/dockerclient.cc [hunk 2]
```diff
@@ -50,4 +51,29 @@ std::tuple<bool, std::string> DockerClient::getContainerState(const Json::Value&
   return {false, ""};
 }
 
+Json::Value DockerClient::getEngineInfo() {
+  Json::Value info;
+  std::string cmd;
+  if (!http_client_) {
```
[the remark this answers is not in the document]
Comment (mike-sul) [reply src/docker/dockerclient.cc:57]: Jorge thinks that using `curl` is preferable compared to the custom http client, since LmP  doesn't include curl by default we end up having both :).

File: src/docker/restorableappengine.cc [hunk 3]
```diff
@@ -54,6 +71,9 @@ bool RestorableAppEngine::fetch(const App& app) {
     res = true;
   } catch (const std::exception& exc) {
     LOG_ERROR << "failed to fetch App; app: " + app.name + "; uri: " + app.uri + "; err: " + exc.what();
```
[the remark this answers is not in the document]
Comment (mike-sul) [reply src/docker/restorableappengine.cc:73]: It makes sense, this is just a topic for another task and PR though.

File: src/docker/restorableappengine.cc [hunk 4]
```diff
@@ -54,6 +71,9 @@ bool RestorableAppEngine::fetch(const App& app) {
...
+    if (boost::filesystem::exists(app_dir)) {
+      boost::filesystem::remove_all(app_dir);
+    }
```
[the remark this answers is not in the document]
Comment (mike-sul) [reply src/docker/restorableappengine.cc:76]: It looks like I came across some subtle issue related to checking if app needs to be (re-)downloaded, hence I put it here as a quick fix. But, I think I will remove it from the given PR and address the issue in another one if there is indeed some issue.

File: src/docker/restorableappengine.cc [hunk 5]
```diff
@@ -276,6 +277,21 @@ void RestorableAppEngine::pullApp(const Uri& uri, const boost::filesystem::path&
...
   const auto archive_full_path{app_dir / (HashedDigest(manifest.archiveDigest()).hash() + Manifest::ArchiveExt)};
 
+  boost::system::error_code ec;
+  boost::filesystem::space_info fs_storage_info{boost::filesystem::space(app_dir, ec)};
+  if (ec.failed()) {
+    LOG_WARNING << "Failed to get an available storage size: " << ec.message();
+  } else {
+    // assume that an extracted files total size is up to 10x larger than the archive size
+    // 80% is a storage space watermark, we don't want to fill a storage volume above it
+    auto need_storage = manifest.archiveSize() * 10;
```
[the remark this answers is not in the document]
Comment (mike-sul) [reply src/docker/restorableappengine.cc:307]: Do you mean the watermark or the compression ration assumption?
I suppose we can do it, but a user can shoot himself in the foot by playing with knob :). I think we can have such param but make sure that we limit  a value range that a user can define.

File: src/docker/restorableappengine.cc [hunk 6]
```diff
@@ -299,6 +319,87 @@ void RestorableAppEngine::pullApp(const Uri& uri, const boost::filesystem::path&
        boost::process::start_dir = app_dir);
 }
 
+void RestorableAppEngine::checkAppUpdateSize(const Uri& uri, const boost::filesystem::path& app_dir) const {
+  const Manifest manifest{Utils::parseJSONFile(app_dir / Manifest::Filename)};
+  const Json::Value engine_info{docker_client_->engineInfo()};
+  if (!(engine_info.isObject() && engine_info.isMember("Arch"))) {
+    LOG_WARNING << "Failed to get an info about a system architecture";
+    return;
+  }
```
[the remark this answers is not in the document]
Comment (mike-sul) [reply src/docker/restorableappengine.cc:328]: docker_client_ instance inquires the docker daemon once to get this info and then stores it its context, so both  engineInfo() and engine_info.isMember("Arch") are merely getters/accessors.
Passing the arch value from pacman down to the engine might require more code than the current approach, and it really is not needed by pacman nor by any other app engine types.
Also, we will need to write if (arch.empty() /* not defined*/)  anyway to make code defensive.

As trade off, we can get an arch value at RestorableAppEngine ctor and keep it in its context, or do the same in the dockerclient scope - get an arch value at DockerClient ctor store in its context as string and add getter to its interface, so no isObject() && engine_info.isMember are required just `f (arch.empty() `.

File: src/docker/restorableappengine.cc (hunk 1 above)
Comment (mike-sul) [reply src/docker/restorableappengine.cc:333]: Not sure. Even after compose-publish is updated there can be Targets that are not installed on devices and such that refer to Apps published by older version of compose-publish. IMO, it's quite valid use-case?

File: src/docker/dockerclient.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (doanac) [reply src/docker/dockerclient.cc:57]: is there an actual code path where this is null? If not - let's remove it all eventually. its too much mental load reading that everywhere.

File: src/docker/restorableappengine.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (doanac) [reply src/docker/restorableappengine.cc:76]: ah - something might be corrupt and we'd not re-download. seems reasonable.

File: src/docker/restorableappengine.cc (hunk 1 above)
Comment (doanac) [reply src/docker/restorableappengine.cc:333]: i guess its something we can adjust later once people are all updated.

File: src/docker/restorableappengine.cc (hunk 5 above)
[the remark this answers is not in the document]
Comment (doanac) [reply src/docker/restorableappengine.cc:307]: i was thinking more the watermark.

File: src/docker/restorableappengine.cc (hunk 6 above)
[the remark this answers is not in the document]
Comment (doanac) [reply src/docker/restorableappengine.cc:328]: i misread. thought this was calling GetEngineInfo.

[the remark this answers is not in the document]
Comment (mike-sul) [reply src/docker/restorableappengine.cc:328]: I am adding the arch value to the  DockerClient ctor, imo, it makes code cleaner as well as the arch value we are looking for must be compliant to the docker architecture naming (e.g. not aarch64 or something else but arm64) so it is kind of natural that the docker arch value is an attribute of DockerClient instance.

File: src/docker/restorableappengine.cc (hunk 5 above)
[the remark this answers is not in the document]
Comment (mike-sul) [reply src/docker/restorableappengine.cc:307]: now we need to come with the param name :). app_store_watermark and allowed range [10, 99] ?

[the remark this answers is not in the document]
Comment (doanac) [reply src/docker/restorableappengine.cc:307]: i'm almost ready to mark this as a "TODO". this might require a little thought and consideration?

File: src/docker/dockerclient.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (mike-sul) [reply src/docker/dockerclient.cc:57]: ok, I'll do it another PR.

File: src/docker/restorableappengine.cc (hunk 5 above)
[the remark this answers is not in the document]
Comment (mike-sul) [reply src/docker/restorableappengine.cc:307]: ok, will create a jira ticket.

File: src/docker/restorableappengine.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (mike-sul) [reply src/docker/restorableappengine.cc:76]: Actually, it's needed for the given feature. For example, if checkAppUpdateSize fails then App's content (App manifest, archive and manifests of images) is actually downloaded so, the follow up check isAppFetched() says yes, the App is fetched while actually it's fetched partially, so I thought that it would be OK and more correct to remove the App content (basically just manifests).
Ideally, we need to improve `isAppFetched` so it checks not just manifests but the blob store (layers) too or leave it as it is and add another method areImagesPulled() and then add it to `RestorableAppEngine::isFetched()`.

I suggest to leave this App's manifests removal and then I will improve isFetched()/isAppFetched in another PR?

**Comment by mike-sul** [comment]
> i think we are good now, right?

please, check the latest update (force push).

**Review by doanac — APPROVED** [review]
